distributed/worker.py

```diff
@@ -2661,16 +2800,15 @@ def release_key(
         reason: Optional[str] = None,
         report: bool = True,
     ):
+        recommendations = {}
         try:
             if self.validate:
                 assert not isinstance(key, TaskState)
-            ts = self.tasks.get(key, None)
-            # If the scheduler holds a reference which is usually the
-            # case when it instructed the task to be computed here or if
-            # data was scattered we must not release it unless the
-            # scheduler allow us to. See also handle_delete_data and
-            if ts is None or ts.scheduler_holds_ref:
-                return
+            ts = self.tasks[key]
+            # needed for legacy notification support
+            state_before = ts.state
+            ts.state = "released"
```

Comment:
The legacy notification support I am referring to is the WorkerPlugin system which has a hook for `release_key`. I don't think this is sensible after this refactoring anymore and believe it should be deprecated.

----

We do have two special states `executing` and `flight` which we are not allowed to simply leave since they are coupled to an "external" system. Executing tasks are coupled to some executor (typically ThreadPool but that's not a given) while flight tasks are relying on a coroutine to be finished. Let's take the example of an executing task.

Previously, if a task was executed and there was a stimulus to release the key [1] the task would be removed from `self.tasks`. Once the ThreadPool finishes with the task execution it would check if the key existed in `self.tasks` and decide what to do based on this information. If it's not part of `self.tasks` it would be safe to simply forget. Else, transition to `memory`.
That's simple. What happens if the key is released *and* rescheduled during the execution time window? The task will be part of `self.tasks` again but could literally be in any other state. For instance, it is in state `fetch` which leads to the frequently reported `("fetch", "memory")` `KeyError`. However, it could also be in the state `flight` and we'd have a coroutine already running. While `("flight", "memory")` is implemented, it is perceived as the transition of a task just arriving from another worker, not a task just done executing.
The problem with this is that we cannot perform the necessary cleanup/state exit action if we do not know what the actual state is. This could even lead to some major inconsistencies if there are actually two or three concurrent states for the same task.

Therefore, if some stimulus generates the _recommendation_ to _release_ a key, it will trigger the `(start, "released")` transition. In case of executing, this transition will actually not transition to `released` but rather in an intermediate state `cancelled` which blocks any further transitioning until the `execute` is `done`. This is more or less the concept behind the _recommendations_, as I understand them, since the transition will not guarantee the final state to follow that recommendation. 
Using this system we can implement the _rule_: "`flight` and `executing` must not be exited until their executor finished".

> This is the only place where "released" is explicitly set as the state

I would hope that all transitions which actually intend to release the key go through this method. This method is intended to reset the state to a neutral, released state. 


[1] There are many examples why this could happen.
* User cancelled the keys
* Partial graph rescheduling due to missing data, e.g. because a worker briefly disconnected.
* Scheduler reschedules the dependency because the only worker with a known replica was lost and *this* worker wasn't able to notify the scheduler about his copy, yet.
* Construct any kind of example with slow, asymetric (read/write perf, serialization, etc.) or disconnected network and you'll get some weird race conditions, I promise. Therefore, at this point I do no longer ask the question "why does this happen" but rather accept that it is happening regardless of why and how.